fedbiomed/node/secagg.py

```diff
@@ -316,48 +309,128 @@ def __init__(
+        self._secagg_manager = DHManager
+        self._grpc_client = grpc_client
+        self._pending_requests = pending_requests
 
-        # create a (currently dummy) context if it does not exist yet
-        time.sleep(3)
-        context = {
-            'biprime': int(random.randrange(10**12)),   # dummy biprime
-            'max_keysize': 0                            # prevent using the dummy biprime for real
-        }
-        logger.info("Not implemented yet, PUT SECAGG BIPRIME GENERATION PAYLOAD HERE, "
-                    f"secagg_id='{self._secagg_id}'")
+        if not self._experiment_id or not isinstance(self._experiment_id, str):
+            errmess = f'{ErrorNumbers.FB318.value}: bad parameter `experiment_id` must be a non empty string'
+            logger.error(errmess)
+            raise FedbiomedSecaggError(errmess)
 
-        # Currently, all biprimes can be used by all sets of parties.
-        # TODO: add a mode where biprime is restricted for `self._parties`
-        self._secagg_manager.add(self._secagg_id, None, context)
+    def _setup_specific(self) -> None:
+        """Service function for setting up the Diffie Hellman secagg context element.
+        """
+        # we know len(parties) >= 3 so len(other_nodes) >= 1
+        other_nodes = [ e for e in self._parties[1:] if e != environ['NODE_ID'] ]
+
+        # Key exchange with other nodes
+        # TODO: replace `dummy` with real payload
+
+        other_nodes_messages = []
+        for node in other_nodes:
+            other_nodes_messages += [
+                NodeToNodeMessages.format_outgoing_message({
+                    'request_id': REQUEST_PREFIX + str(uuid.uuid4()),
+                    'node_id': environ['NODE_ID'],
+                    'dest_node_id': node,
+                    'dummy': f"KEY REQUEST INNER from {environ['NODE_ID']}",
+                    'secagg_id': self._secagg_id,
+                    'command': 'key-request'
+                })
+            ]
+
+        logger.debug(f'Sending Diffie-Hellman setup for {self._secagg_id} to nodes: {other_nodes}')
+        listener_id = send_overlay_message(
+            self._grpc_client,
+            self._pending_requests,
+            self._researcher_id,
+            other_nodes,
+            other_nodes_messages,
+        )
+        all_received, messages = self._pending_requests.wait(listener_id, TIMEOUT_NODE_TO_NODE_REQUEST)
+
+        logger.debug(f"Completed Diffie-Hellmann setup with success={all_received} "
+                     f"node_id='{environ['NODE_ID']}' secagg_id='{self._secagg_id}")
+        if not all_received:
+            nodes_no_answer = set(other_nodes) - set([m['node_id'] for m in messages])
+            msg = f"{ErrorNumbers.FB318.value}: Some nodes did not answer during Diffie Hellman secagg " \
+                  f"context setup: {nodes_no_answer}"
+            logger.error(msg)
+            raise FedbiomedSecaggError(msg)
```

Comment:
Hi @mvesin,
`logger.error` can be called where this exception is caught in the end. For example, before sending reply in `setup` function, because there are also other exception may be raised and caught in the same place.

Reply:
Hi @srcansiz
Thank you for the review.

I agree this is not needed. We can completely discard this `logger` entry.

The logging you propose in `SecaggBaseSetup.setup()` already exists (with `debug` instead of `error`), so my `logger` entry in `_setup_specific` is really duplicate.
An `error` entry is also launched later in the `self._create_secagg_reply()` called from `setup()`

```
2024-06-25 08:56:48,349 fedbiomed DEBUG - Completed Diffie-Hellmann setup with success=False node_id='NODE_e2853014-c202-4a62-83b5-2662a4e8495d' secagg_id='secagg_178c1ca5-cc4a-4d06-8fe2-f2ad719f032c
2024-06-25 08:56:48,349 fedbiomed ERROR - FB318: Secure aggregation setup error: Some nodes did not answer during Diffie Hellman secagg context setup: {'NODE_9ff33d22-a893-4649-9535-ebf465a5221f'}
2024-06-25 08:56:48,350 fedbiomed DEBUG - FB318: Secure aggregation setup error: Some nodes did not answer during Diffie Hellman secagg context setup: {'NODE_9ff33d22-a893-4649-9535-ebf465a5221f'}
2024-06-25 08:56:48,350 fedbiomed ERROR - Can not setup secure aggregation context on node for secagg_178c1ca5-cc4a-4d06-8fe2-f2ad719f032c.
```